app/lib/actions/observable_action.rb

```diff
@@ -0,0 +1,72 @@
+module Actions
+  # Examples:
+
+  # # Action A which emits an event when it successfully finishes.
+  # class A
+  #   include ::Actions::ObservableAction
+  #   # ... rest ...
+  # end
+
+  # # Action B which emits an event when it successfully finishes or fails.
+  # class B
+  #   include ::Actions::ObservableAction
+  #
+  #   execution_plan_hooks.use :emit_event_failure, :on => [:failure]
+  #
+  #   def self.event_names
+  #     super + [event_name_base + '_' + event_name_suffix(:failure)]
+  #   end
+  #
+  #   def emit_event_failure(plan)
+  #     emit_event(plan, :failure)
+  #   end
+  #   # ... rest ...
+  # end
+  module ObservableAction
+    module ClassMethods
+      def event_name_suffix(hook)
+        case hook
+        when :success
+          'succeeded'
+        when :failure
+          'failed'
+        else
+          hook
+        end
+      end
+
+      def event_names
+        [event_name_base + '_' + event_name_suffix(:success)]
+      end
+
+      def namespaced_event_names
+        event_names.map { |e| ::Foreman::Observable.event_name_for(e) }
+      end
+
+      def event_name_base
+        to_s.underscore.tr('/', '.')
+      end
+    end
+
+    def self.included(base)
+      base.extend ClassMethods
+      base.include ::Foreman::Observable
+      base.execution_plan_hooks.use :emit_event, :on => :success
```

Comment:
If se use the hook only on success, will the `*_failed` events ever be emitted?

Reply:
Not by default, We decided to emit events on success by default, however action authors can decide to emit events also on different execution plan state transitions. Some parts of the code are there just in case someone does that